### ISSUE TITLE
False positive missing-argument errors on hypothesis composite strategies

### ISSUE THREAD

**Issue by lukasbindreiter** [astral-sh/ty#2759]
### Summary

Hi,

the latest release of ty `0.0.15` has a change that results in lots of false-positive `missing-argument` errors in our code base.

It's related to usages of [hypothesis.composite](https://hypothesis.readthedocs.io/en/latest/reference/strategies.html#hypothesis.strategies.composite)

Here is a quick example

```python
from hypothesis import given
from hypothesis.strategies import DrawFn, composite, integers, none

@composite
def optional_ints(draw: DrawFn) -> int | None:
    return draw(integers() | none())


# below line works in ty 0.0.14, but results in an missing-argument error in 0.0.15
@given(optional_ints())
def test_something(x: int | None):
    if x is None:
        ...
    pass
```

Type checking this results in the following false-positive error:

```
> uv run ty version
ty 0.0.15 (eb43dff06 2026-02-04)

> uv run ty check
error[missing-argument]: No arguments provided for required parameters `*args`, `**kwargs`
  --> main.py:10:8
   |
 9 | # below line works in ty 0.0.14, but results in an missing-argument error in 0.0.15
10 | @given(optional_ints())
   |        ^^^^^^^^^^^^^^^
11 | def test_something(x: int | None):
12 |     if x is None:
   |
info: These arguments are required because `ParamSpec` `P'return` could represent any set of parameters at runtime
info: rule `missing-argument` is enabled by default

Found 1 diagnostic
```

For comparison, it worked fine before the latest release:

```
> uv run ty version
ty 0.0.14 (16597f5e2 2026-01-26)

> uv run ty check
All checks passed!
```

My full `pyproject.toml` uv venv looks like this:

<details>

<summary>`pyproject.toml`</summary>

```toml
[project]
name = "ty-test"
version = "0.1.0"
description = "ty false-positive error demo"
readme = "README.md"
requires-python = ">=3.13"
dependencies = [
    "hypothesis>=6.151.5",
    "ty>=0.0.15",
]
```
</details>

And for reference, `hypothesis.composite` has the following type definition ([source](https://github.com/HypothesisWorks/hypothesis/blob/hypothesis-python-6.151.5/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L2058))

```python
P = ParamSpec("P")

def composite(
    f: Callable[Concatenate[DrawFn, P], Ex],
) -> Callable[P, SearchStrategy[Ex]]:
    return _composite(f)
```

### Version

`0.0.15`

**Comment by sharkdp** [comment]
@dhruvmanila I haven't investigated further, but the symptoms look similar to #2724.

**Comment by carljm** [comment]
Here's a simplified standalone repro:

```py
from __future__ import annotations
from typing import Callable, Concatenate, ParamSpec, TypeVar

P = ParamSpec("P")
Ex = TypeVar("Ex")

class DrawFn:
    pass

class SearchStrategy:
    pass

def composite(
    f: Callable[Concatenate[DrawFn, P], Ex],
) -> Callable[P, SearchStrategy]:
    ...

@composite
def optional_ints(draw: DrawFn) -> int | None:
    return None

reveal_type(optional_ints)
x = optional_ints()
```

The proximate cause is https://github.com/astral-sh/ruff/pull/22954, but the root cause is lack of `Concatenate` support. Because of the lack of `Concatenate` support, we wrongly see `P` as "unused" in the arguments of `composite`, so wrongly treat it as generic in the returned callable. In order to fix this, I think we either need to add full `Concatenate` support, or at least add enough of a representation of it that we can see that `P` is used in the arguments.

**Comment by Hugo-Polloli** [comment]
I started working on a fix for this yesterday. It’s not full `Concatenate` support yet, but I’ve thought it in a way that should compose cleanly with broader `Concatenate` support

**Comment by AlexWaygood** [comment]
> I started working on a fix for this yesterday. It’s not full `Concatenate` support yet, but I’ve thought it in a way that should compose cleanly with broader `Concatenate` support

(note that we do already have a contributor PR open for `Concatenate` support — though I haven't looked at it at all yet! https://github.com/astral-sh/ruff/pull/23119)

**Comment by Hugo-Polloli** [comment]
> > I started working on a fix for this yesterday. It’s not full `Concatenate` support yet, but I’ve thought it in a way that should compose cleanly with broader `Concatenate` support
> 
> (note that we do already have a contributor PR open for `Concatenate` support — though I haven't looked at it at all yet! [astral-sh/ruff#23119](https://github.com/astral-sh/ruff/pull/23119))

Oh I missed that haha
I'll check the PR and see if there any of my insights might be worth sharing there! :)

**Comment by carljm** [comment]
@dhruvmanila It seems like just adding Concatenate support is the best option here, so assigning this to you since you're working on that.

**Comment by dhruvmanila** [comment]
ty doesn't raise this error now that we have `Concatenate` support (https://github.com/astral-sh/ruff/pull/23689) but the type of `test_something` is `(...) -> None` instead of `(int | None) -> None` which is same as all other type checkers (Pyright, pyrefly, mypy) and is expected given how the `hypothesis.given` function is typed:

```py
@overload
def given(
    _: EllipsisType, /
) -> Callable[
    [Callable[..., Coroutine[Any, Any, None] | None]], Callable[[], None]
]:  # pragma: no cover
    ...


@overload
def given(
    *_given_arguments: SearchStrategy[Any],
) -> Callable[
    [Callable[..., Coroutine[Any, Any, None] | None]], Callable[..., None]
]:  # pragma: no cover
    ...


@overload
def given(
    **_given_kwargs: SearchStrategy[Any] | EllipsisType,
) -> Callable[
    [Callable[..., Coroutine[Any, Any, None] | None]], Callable[..., None]
]:  # pragma: no cover
    ...
```

Here's the original example with added `reveal_type`s:

```py
from typing import reveal_type

from hypothesis import given
from hypothesis.strategies import DrawFn, composite, integers, none


@composite
def optional_ints(draw: DrawFn) -> int | None:
    return draw(integers() | none())


reveal_type(optional_ints)  # () -> SearchStrategy[int | None]
reveal_type(optional_ints())  # SearchStrategy[int | None]
# ((...) -> Coroutine[Any, Any, None] | None, /) -> ((...) -> None)
reveal_type(given(optional_ints()))


@given(optional_ints())
def test_something(x: int | None):
    if x is None:
        ...
    pass


# (...) -> None
reveal_type(test_something)
```
